staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go

```diff
@@ -513,3 +515,135 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
+	closing              *atomic.Bool
+}
+
+// NewDynamicTransformers initializes DynamicTransformers when api-server starts.
+func NewDynamicTransformers(
+	transformerOverrides map[schema.GroupResource]value.Transformer,
+	closeTransformers context.CancelFunc,
+	stopCh <-chan struct{},
+) *DynamicTransformers {
+	dynamicTransformers := &DynamicTransformers{
+		transformTracker: &atomic.Value{},
+		stopCh:           stopCh,
+	}
+
+	tracker := &transformTracker{
+		transformerOverrides: transformerOverrides,
+		closeTransformers:    closeTransformers,
+		inFlight:             &atomic.Int64{},
+		closing:              &atomic.Bool{},
+	}
+	dynamicTransformers.transformTracker.Store(tracker)
+
+	return dynamicTransformers
+}
+
+// TransformerForResource returns the transformer for the given resource.
+func (d *DynamicTransformers) TransformerForResource(resource schema.GroupResource) value.Transformer {
+	return &resourceTransformer{
+		resource:         resource,
+		transformTracker: d.transformTracker,
+	}
+}
+
+// Set sets the transformer overrides. This method is not go routine safe and must only be called by the same, single caller throughout the lifetime of this object.
+func (d *DynamicTransformers) Set(
+	transformerOverrides map[schema.GroupResource]value.Transformer,
+	closeTransformers context.CancelFunc,
+) {
+	// store new values
+	newTransformTracker := &transformTracker{
+		transformerOverrides: transformerOverrides,
+		closeTransformers:    closeTransformers,
+		inFlight:             &atomic.Int64{},
+		closing:              &atomic.Bool{},
+	}
+
+	oldTransformTracker := d.transformTracker.Swap(newTransformTracker).(*transformTracker)
+
+	oldTransformTracker.closing.Store(true)
+	if oldTransformTracker.inFlight.Load() == 0 {
+		oldTransformTracker.closeTransformers()
+	}
+}
+
+var _ value.Transformer = &resourceTransformer{}
+
+type resourceTransformer struct {
+	resource         schema.GroupResource
+	transformTracker *atomic.Value
+}
+
+func (r *resourceTransformer) TransformFromStorage(ctx context.Context, data []byte, dataCtx value.Context) ([]byte, bool, error) {
+	// this might cause race condition between two goroutines calling this method at the same time.
+	// due to this old transformer might be closed before they are done serving the request.
+	// ToDo: come up with a better solution to make sure the transformer is not closed when in use.
+	tracker := r.transformTracker.Load().(*transformTracker)
+
+	tracker.inFlight.Add(1)
+	defer func() {
+		if tracker.inFlight.Add(-1) == 0 && tracker.closing.Load() {
+			tracker.closeTransformers()
+		}
+	}()
+
+	if tracker.closing.Load() && tracker.inFlight.Load() == 0 {
+		tracker.closeTransformers()
+		return nil, false, fmt.Errorf("transformer is closing")
+	}
+
+	transformer := tracker.transformerOverrides[r.resource]
+	if transformer == nil {
+		transformer = identity.NewEncryptCheckTransformer()
+	}
+
+	return transformer.TransformFromStorage(ctx, data, dataCtx)
+}
```

Comment:
@liggitt any ideas on how to make this not racey without adding a lock?  It is trivial to fix with a lock, but I know we would not be okay with a regular mutex here, though maybe a read write mutex would be okay (i.e. `SafeWaitGroup`).  One questionable idea I had was to drop all the in flight tracking and simply sleep for a minute at the end of `.Set` before calling `oldTransformTracker.closeTransformers()`...

Reply:
Hi @liggitt, would love to hear your thoughts on this. Thanks.

Reply:
updates as per feedback.